### ISSUE TITLE
[Impeller] Impeller RRect blur is slower than Skia's cached blur shadows.

### ISSUE THREAD

**Issue by voxelbee** [flutter/flutter#148496]
### Steps to reproduce

1. Create a new project use the code sample bellow and make sure you are running using impeller backend.
2. Scroll in the list and there will be a large amount of jank during scrolling.
3. Make sure to use a custom painter with a shader and layers as show in the example code. This makes the jank much more visible.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: ListView.builder(
        itemCount: 100,
        itemBuilder: (context, index) {
          return ListItemWidget(
            key: ValueKey<int>(index),
            index: index,
          );
        },
      ),
    );
  }
}

class ListItemWidget extends StatelessWidget {
  const ListItemWidget({required this.index, super.key});

  final int index;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(8),
      child: Painter(
        child: Card(
          child: ListTile(
            title: Text('Item $index'),
          ),
        ),
      ),
    );
  }
}

class Painter extends StatelessWidget {
  const Painter({required this.child, super.key});

  final Widget child;

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      painter: _CustomPainter(),
      child: child,
    );
  }
}

class _CustomPainter extends CustomPainter {
  _CustomPainter();

  static const List<Offset> bubbles = [
    Offset(0.2, 0.2),
    Offset(0.8, 0.8),
    Offset(0.5, 0.5),
  ];

  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint()
      ..style = PaintingStyle.fill
      ..filterQuality = FilterQuality.low
      ..maskFilter = const MaskFilter.blur(BlurStyle.normal, 35);

    final rect = (Offset.zero & size).inflate(40);

    canvas.saveLayer(rect, Paint());
    for (final bubble in bubbles) {
      final position = Offset(
        size.width * (1 - (bubble.dx * 0.4)),
        size.height * bubble.dy,
      );

      canvas.drawCircle(
        position,
        50,
        paint..color = Colors.red,
      );
    }

    canvas
      ..drawRect(
        rect,
        Paint()
          ..shader = const LinearGradient(
            begin: Alignment.topCenter,
            end: Alignment.bottomCenter,
            colors: [
              Colors.transparent,
              Colors.white,
              Colors.white,
              Colors.transparent,
            ],
            stops: [0.2, 0.3, 0.7, 0.8],
          ).createShader(rect)
          ..blendMode = BlendMode.dstIn,
      )
      ..drawRect(
        rect,
        Paint()
          ..shader = const LinearGradient(
            colors: [
              Colors.transparent,
              Colors.white,
              Colors.white,
              Colors.transparent,
            ],
            stops: [0.06, 0.1, 0.85, 0.94],
          ).createShader(rect)
          ..blendMode = BlendMode.dstIn,
      )
      ..restore();
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return false;
  }
}
```

</details>

### Performance profiling on master channel

- [X] The issue still persists on the master channel

### Timeline Traces

<details open><summary>Timeline Traces JSON</summary>

[dart_devtools_2024-05-16_17 42 26.249.json.zip](https://github.com/flutter/flutter/files/15338430/dart_devtools_2024-05-16_17.42.26.249.json.zip)

<img width="2192" alt="Screenshot 2024-05-16 at 4 46 42 PM" src="https://github.com/flutter/flutter/assets/11301647/f43f4a09-11ac-4371-944e-e8d3a75a165f">

</details>

### Video demonstration

<details open>
<summary>Video demonstration</summary>

[Upload media here]

</details>

### What target platforms are you seeing this bug on?

iOS

### OS/Browser name and version | Device information

iOS 17.4.1
IPhone 13 Pro Max 

### Does the problem occur on emulator/simulator as well as on physical devices?

Unknown

### Is the problem only reproducible with Impeller?

Yes

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel master, 3.22.0-35.0.pre.35, on macOS 14.0 23A344 darwin-arm64, locale en-US)
    • Flutter version 3.22.0-35.0.pre.35 on channel master at /opt/homebrew/Caskroom/flutter/3.19.2/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision c719f03ded (18 minutes ago), 2024-05-16 18:12:33 +0200
    • Engine revision 460df6caef
    • Dart version 3.5.0 (build 3.5.0-160.0.dev)
    • DevTools version 2.36.0-dev.10

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at /Users/peytonhammersley/Library/Android/sdk
    • Platform android-34, build-tools 33.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.15.2

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.87.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.86.0

[✓] VS Code (version 1.90.0-insider)
    • VS Code at /Applications/Visual Studio Code - Insiders.app/Contents
    • Flutter extension version 3.88.0

[✓] Connected device (2 available)
    • Peyton’s iPhone (mobile)        • 00008110-000A1C2811A3801E • ios    • iOS 17.4.1 21E236
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad     • darwin • macOS 14.0 23A344 darwin-arm64

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by jonahwilliams** [comment]
Debugging this a bit, the cost is evenly split between the rrect blur and the linear gradient shader. The ALU is limiting operations, which I'd expect for rrect blur but gradient is a bit surprising (well, not that surprising). So if we need an example of where some elbow grease is needed for rrect blur or gradient this is probably it.

![image](https://github.com/flutter/flutter/assets/8975114/56ddb3c8-289b-4a0a-86cd-3e55879d9476)

**Comment by mark8044** [comment]
Ive been noticing increasing jank in list views, which was only very minor and so subtle as not noticeable in 3.10.x until now its more noticeable and mildly annoying in 3.22.  I do have alot of gradients and blurs and images inside my listview

**Comment by jonahwilliams** [comment]
@mark8044  Without more information (i.e. a full bug report) we won't be able to say if that is the same issue as here.

**Comment by gaaclarke** [comment]
There is a fair bit of overdraw happening:
![overdraw](https://github.com/flutter/flutter/assets/30870216/91b65185-c2b2-484f-bba0-fa92b7310c23)

The team can work to make this faster but if this is representative of what you are really drawing I think the drawing could be simplified by drawing one transformed oval (instead of overlapping circles to make an oval) and then doing one dstin blend to clip it.  The first one seems superfluous, in this example at least.

**Comment by jonahwilliams** [comment]
@gaaclarke if you look at the limiters though, its ALU ops and not memory writes. So while reducing overdraw will help its not the cause of the slowness w.r.t. skia

**Comment by gaaclarke** [comment]
> @gaaclarke if you look at the limiters though, its ALU ops and not memory writes. So while reducing overdraw will help its not the cause of the slowness w.r.t. skia

Yea I saw that.  Simplifying this to one oval and one gradient `dstIn` rect will reduce the math.  It's not so much the overdraw in particular, but drawing 3 things to make one thing via overdraw is less efficient math-wise and memory-wise.

**Comment by voxelbee** [comment]
> The team can work to make this faster but if this is representative of what you are really drawing I think the drawing could be simplified by drawing one transformed oval (instead of overlapping circles to make an oval) and then doing one dstin blend to clip it. The first one seems superfluous, in this example at least.

Yea well my actual code uses different colors for each of the circles so I was rendering them each separate. I guess I could optimize this a bit so that it renders one with transformations and a color blend instead? However, this wont still fix the issue that this is still incredibly slow. This doesn't seem to be the case when rendering the exact same UI with Skia.

**Comment by jonahwilliams** [comment]
Right, there may be ways to optimize the UI but we should still make a reasonable effort to be at least as fast as Skia.

**Comment by gaaclarke** [comment]
I landed one patch that reduces the arithmetic cycles for the linear gradient.  It does also bump the required number of registers slightly though.  Once it has merged into the framework it would be nice if the OP could give it a look.

I've spent some time looking into rrect_blur.  The majority of the time comes from special functions (exp and sqrt).  For the regular blur I was able to pre-calculate the gaussian blur coefficients on the CPU.  It's difficult to do here since we are manipulating the blur based on the x and y values of the position, but I think it should be possible.  I also briefly looked at our exp and sqrt usage.
 - One exp is used for gaussian distribution calculation (in a loop)
 - One exp is used for gaussian integration estimation.  So even though exp is slow it is already being used as an optimization to avoid full integration.
 - The sqrt is for finding distances for corners.  I tried to look to see if we could perform the math squared space to avoid the sqrt.  The distance is used to calculate the range over which we will integrate, which if we have a real gaussian distribution, overshooting it a bit might not be a problem if we wanted to estimage the sqrt.  But since we are using an estimation for the integration I'm not so confident.
 
Jonah reminded me we should see where we can take advantage of lower precision.  So I'm keeping an eye out for that.  I might end up having to move the code to the cpu so I can get a better idea of what sort of numbers we're working with.

TLDR linear gradient should be faster, rrect_blur is a lot trickier to optimize.

**Comment by gaaclarke** [comment]
I transitioned rrect_blur to c++ so it can be examined for optimizations on the cpu: https://gist.github.com/gaaclarke/36ed9fad08f2a34a9230a61b328ea3cc

**Comment by gaaclarke** [comment]
I don't know if it's a fluke of my test, but the majority of the time `sqrt` in rrect_blur isn't used since we have an argument that is 1 or negative (which we throw away).  I threw up a PR that will take advantage of that.  The weird thing though is that when it isn't thrown away, only one ever argument is ever sent to `sqrt`.  I'm not sure if we can pre-calculate that.  It wasn't obvious that we could when i was staring at this.

**Comment by voxelbee** [comment]
> I landed one patch that reduces the arithmetic cycles for the linear gradient. It does also bump the required number of registers slightly though. Once it has merged into the framework it would be nice if the OP could give it a look.

I just tested out the change it looks a bit better now then it was before the changes but it is still very far off from the performance with Skia. In Skia there isn't any frame spikes when scrolling and it runs really smooth. However, in Impeller there is still really visible frame spikes and jitter when scrolling. Also just posting `flutter doctor` output to check that I'm running the right version with the changes.

```bash
[✓] Flutter (Channel master, 3.23.0-9.0.pre.3, on macOS 14.5 23F79 darwin-arm64, locale en-US)
    • Flutter version 3.23.0-9.0.pre.3 on channel master at /opt/homebrew/Caskroom/flutter/3.19.2/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a1a33e63b9 (6 hours ago), 2024-05-28 03:56:25 -0400
    • Engine revision b1751088c7
    • Dart version 3.5.0 (build 3.5.0-191.0.dev)
    • DevTools version 2.36.0-dev.10
```

**Comment by gaaclarke** [comment]
Thanks for checking, that version of flutter would have had the optimization at https://github.com/flutter/engine/pull/53007.  Possible further optimizations:
1) Conditionally doing `sqrt` in rrect_blur - we have a draft PR but the performance of gpu's surrounding branching is complicated
1) Eliminating the `sqrt` in rrect_blur - the cpu port of the shader seems to only call `sqrt` on a couple of numbers, seems like we could potentially pre-calculate that.
1) Moving the gaussian calculation (`exp`) to the cpu in rrect_blur - this is how I made the gaussian blur faster.  I think it may be possible here too.  It's a bit tricky since the fragment shader isn't reading from the whole curve like the gaussian blur.
1) Dropping fidelity - either in terms of floating point precision or render size.

Beyond that I'm not sure.  The linear gradient could be optimized for the number of colors, but more variants creates performance problems elsewhere.

**Comment by jonahwilliams** [comment]
We can do much, much better with linear gradients: https://github.com/flutter/flutter/issues/148651

**Comment by jonahwilliams** [comment]
Numbers with https://github.com/flutter/engine/pull/53166:

Skia: 3.5ms
Impeller ToT: 20ms
Impeller Patched: 11ms

So fixing gradients shaves off about half the time. To get the rest, we need to improve the performance of the rrect blur.

**Comment by Livinglist** [comment]
Don’t know if this is related but I think my app got more janks in 3.22 than in 3.19 on scrolling.

**Comment by jonahwilliams** [comment]
@Livinglist  its almost certainly not. Please file a bug with details so we can investigate it.

**Comment by jonahwilliams** [comment]
Ahh so for the rrect shader, the difference with Skia is that Skia has a specialized circle case which is much simpler.

**Comment by jonahwilliams** [comment]
@wilpar sorry you need to file an actual bug for us to do anything about this.

**Comment by wilpar** [comment]
Fear not, I plan to ;) ... searching to see if I'm the only one with the problem is usually how I start debugging.

When I ran without impeller, performance was back to normal.

```
flutter run --no-enable-impeller
```

I'll try to cut out a section of the app for a bug report.

**Comment by jonahwilliams** [comment]
I added some more adjustments - recognizing that all of the mask filters are computing the same exact blur. This allows me to dedup all of the blurs into a single compuation, speeding this up to about ~3ms per frame. This is about 0.5 ms better for Skia, so assuming we can land something here we could call it good :)

![image](https://github.com/flutter/flutter/assets/8975114/98d6b891-9b6b-407d-bdf3-582a0152b446)

**Comment by jonahwilliams** [comment]
I haven't gotten back to this in a while. While we landed the changes to improve gradient performance, mask blur performance is still backed up behind other work

**Comment by gaaclarke** [comment]
Depending how we address it, it might also resolve https://github.com/flutter/flutter/issues/152887.  It's worth checking that issue if this one gets closed.

**Comment by gaaclarke** [comment]
I pulled out the gaussian distribution to a texture that we sample.  If anything it is slightly slower: https://github.com/flutter/engine/commit/54b76064bac26a778c432b929cf06dd3bc2e8453

I was hoping that since we are capturing an `exp` and a couple other ops this could be a win.  I'll see if I can improve on it.  Some ideas:
1) Right now it's looping over kSampleCount and sampling multiple times.  Maybe we can prebake that too.
1) Implement Texture1D
1) Switch to kDevicePrivate

**Comment by gaaclarke** [comment]
I went back and looked at the old PR where I added a benchmark for this, looks like we used to measure 12ms/frame.  Now I'm measuring 3-4ms ( https://github.com/flutter/flutter/pull/149261 ).

**Comment by gaaclarke** [comment]
After https://github.com/flutter/engine/pull/55604 rrect blurs are about 60% faster than they were before https://github.com/flutter/engine/pull/55576.

Skia with our benchmark is at 2.5ms, where we are at 4.45ms.  So skia is still faster, although they have a rendering [bug](https://g-issues.skia.org/issues/371571787) which I think they should be able to fix with minimal impact to their performance.  The benchmark is pretty close to a limit of a reasonable amount of rrect blurs one would have on a screen and has some healthy headroom still.

One potential optimization we decided to pass on is creating a specialized variant of the squircle shader for when the exponent used is equal to 2.  [Using the `length` function](https://github.com/flutter/engine/pull/55604#issuecomment-2394748643) instead of implementing our own `powerDistance` function leads to a 15% reducing in the benchmark (or a 24% improvement considering the shader).  The exponent is equal to 2 for circles for sufficiently blurred rrects (the amount of blur depends on the corner radius).

Considering using `length` eliminates the heftiest part of the squircle approach there is probably little left to squeeze from there.  We are getting numbers low enough that I worry if we were to switch to a sampling technique it still may not be as fast as the skia approach because they are working potentially different overhead outside the shader.  If you remove the `powerDistance` function from the squircle implementation there is very little left to it, one `sqrt` and one `length` and a handful of multiplications and subtractions.

**Comment by gaaclarke** [comment]
I looked a bit closer at what skia does.  They have a separate path for circle blur and the rrect blur.  For the circle blur they just create a 1d texture like I was doing in https://github.com/flutter/engine/commit/54b76064bac26a778c432b929cf06dd3bc2e8453 that represents half of the gaussian distribution and then use `length` to calculate where to sample from.  That's the path that is getting hit in this reproduction.  It isn't really rrect_blur.

https://github.com/google/skia/blob/1349ddc074ad616c6660eb29f75dafe18f3714c0/src/gpu/ganesh/GrBlurUtils.cpp#L510

**Comment by jonahwilliams** [comment]
What is the performance gain of specializing the rrect shader on this app? (i.e. if you just hack in the change for length)

**Comment by gaaclarke** [comment]
> What is the performance gain of specializing the rrect shader on this app? (i.e. if you just hack in the change for length)

If we stick with the squircle approach and do `length` it's 24% faster, that's what I mentioned in https://github.com/flutter/engine/pull/55604#issuecomment-2394769634

For the sampler approach I'd have to play with it more.  My sampler approach above was trying to speed up our old algorithm, it would have been smarter to just avoid the old algorithm and do what skia is doing.  I was babystepping to that.

**Comment by jonahwilliams** [comment]
Yes, but on the example app what is the change in ms? because I'm assuming there are other performance differences) I'm less interested in the impact on the synthetic benchmark because ... well its synthetic.

**Comment by gaaclarke** [comment]
> Yes, but on the example app what is the change in ms? because I'm assuming there are other performance differences) I'm less interested in the impact on the synthetic benchmark because ... well its synthetic.

Squircle with circle variant (using `length`) is 10% faster for this reproduction (3.31ms / 3.65ms).

**Comment by jonahwilliams** [comment]
Okay, for now I'd say its not quite worth it. I think we made great progress on this issue, and I'm happy to call it fixed!
